### PR TITLE
Properly mark application/xhtml+xml documents as XML

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14423)

### CODE REVIEW

**Comment by nox** [comment]
```
Tests with unexpected results:
  ▶ Unexpected subtest result in /dom/nodes/Document-getElementsByTagName-xhtml.xhtml:
  └ PASS [expected FAIL] HTML element with uppercase tag name matches in XHTML documents

  ▶ Unexpected subtest result in /dom/nodes/Document-getElementsByTagName-xhtml.xhtml:
  └ PASS [expected FAIL] Element in HTML namespace, no prefix, non-ascii characters in name

  ▶ Unexpected subtest result in /dom/nodes/Document-getElementsByTagName-xhtml.xhtml:
  └ PASS [expected FAIL] Element in HTML namespace, prefix, non-ascii characters in name

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement(undefined) in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement(null) in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("f1oo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo1") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("f\u0bc6") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo\u0bc6") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement(":foo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("f:oo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo:") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("f:o:o") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("f::oo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("f::oo:") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo:0") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo:_") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo:\u0bc6") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo:foo\u0bc6") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("foo\u0bc6:foo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("xml") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("xmlns") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("xmlfoo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("xml:foo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("xmlns:foo") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("xmlfoo:bar") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("svg") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("math") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("FOO") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("mar\u212a") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("\u0130nput") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Document-createElement.html:
  └ PASS [expected FAIL] createElement("\u0131nput") in XHTML document

  ▶ Unexpected subtest result in /dom/nodes/Node-nodeName-xhtml.xhtml:
  └ PASS [expected FAIL] For Element nodes, nodeName should return the same as tagName.

  ▶ Unexpected subtest result in /html/dom/dynamic-markup-insertion/document-write/document.write-01.xhtml:
  └ PASS [expected FAIL] document.write in XHTML

  ▶ Unexpected subtest result in /html/dom/dynamic-markup-insertion/document-writeln/document.writeln-01.xhtml:
  └ PASS [expected FAIL] document.writeln in XHTML
```

File: components/script/dom/element.rs [hunk 1]
```diff
@@ -1800,7 +1832,7 @@ impl ElementMethods for Element {
                                                 ElementCreator::ScriptCreated);
                 Root::upcast(body_elem)
             },
-            _ => context_node.GetParentNode().unwrap()
+            _ => Root::downcast::<Element>(context_node.GetParentNode().unwrap()).unwrap()
```
Comment (Ms2ger) [inline components/script/dom/element.rs]: `GetParentElement().unwrap()`

File: components/script/dom/element.rs [hunk 2]
```diff
@@ -1230,6 +1232,37 @@ impl Element {
...
+        Ok(fragment)
+    }
+
+    pub fn fragment_parsing_context(owner_doc: &Document, element: Option<&Self>) -> Root<Self> {
+        match element {
+            Some(elem) if elem.local_name() != &local_name!("html") || !elem.html_element_in_html_document() => {
+                Root::from_ref(elem)
+            },
+            _ => {
+                Root::upcast(HTMLBodyElement::new(local_name!("body"), None, &owner_doc))
```
Comment (Ms2ger) [inline components/script/dom/element.rs]: No need for the `&`.

File: components/script/dom/node.rs [hunk 3]
```diff
@@ -1811,68 +1809,22 @@ impl Node {
...
-                     // Step 1.
-                     None => ns!(),
-                     // Step 2.
-                     Some(parent) => Node::locate_namespace(parent.upcast(), prefix)
-                 }
+            NodeTypeId::DocumentType | NodeTypeId::DocumentFragment => ns!(),
+            _ => {
+                node.GetParentElement().as_ref()
+                    .map(|elem| elem.locate_namespace(prefix))
+                    .unwrap_or(ns!())
```
Comment (Ms2ger) [inline components/script/dom/node.rs]: `map_or()`?

File: components/script/dom/element.rs [hunk 4]
```diff
@@ -722,6 +723,44 @@ impl Element {
...
+
+        // Steps 3-4.
+        for element in inclusive_ancestor_elements {
+            // Step 1.
+            if element.namespace() != &ns!() && element.prefix().map(|p| &**p) == prefix.as_ref().map(|p| &**p) {
+                return element.namespace().clone();
+            }
+
+            // Step 2.
+            let attr = ref_filter_map(self.attrs(), |attrs| attrs.iter().find(|attr| {
```
Comment (Ms2ger) [inline components/script/dom/element.rs]: Wrap the closure body in a block.

File: components/script/dom/node.rs [hunk 5]
```diff
@@ -1811,68 +1809,22 @@ impl Node {
...
-                    // Step 1.
-                    None => ns!(),
-                    // Step 2.
-                    Some(document_element) => {
-                        Node::locate_namespace(document_element.upcast(), prefix)
-                    }
-                }
+                node.downcast::<Document>().unwrap()
+                    .GetDocumentElement().as_ref()
+                    .map(|elem| elem.locate_namespace(prefix))
```
Comment (Ms2ger) [inline components/script/dom/node.rs]: `map_or()`?

File: components/script/dom/element.rs [hunk 6]
```diff
@@ -1276,12 +1276,8 @@ impl Element {
     pub fn parse_fragment(&self, markup: DOMString) -> Fallible<Root<DocumentFragment>> {
         // Steps 1-2.
         let context_document = document_from_node(self);
-        let new_children = if context_document.is_html_document() {
-            ServoParser::parse_html_fragment(self, markup)
-        } else {
-            // FIXME: XML case
-            unimplemented!()
-        };
+        // TODO: XML case.
```
Comment (Ms2ger) [inline components/script/dom/element.rs]: Check if there is an issue; if not, file one. In either case, make this `FIXME(#NNNN): XML case.`

File: components/script/dom/document.rs [hunk 7]
```diff
@@ -136,7 +136,7 @@ pub enum TouchEventResult {
     Forwarded,
 }
 
-#[derive(JSTraceable, PartialEq, HeapSizeOf)]
+#[derive(Clone, Copy, HeapSizeOf, JSTraceable, PartialEq)]
```
Comment (Ms2ger) [inline components/script/dom/document.rs]: Make it `Debug` while you're here.

**Review by Ms2ger — CHANGES_REQUESTED** [review]

**Review by Ms2ger — APPROVED** [review]

**Pull request merged**